Add pre-commit check in continuous integration

The develop branch failed go-lint.
```
go lint..................................................................Failed
hookid: go-lint

sqlfs/writer.go:8:5: don't use leading k in Go names; var kBufSize should be bufSize
sqlfs/writer.go:55:1: exported method Writer.Close should have comment or be unexported
Found 2 lint suggestions; failing.
server/sqlflowserver.go:1:1: package comment should be of the form "Package server ..."
server/sqlflowserver.go:18:68: exported func NewServer returns unexported type *server.server, which can be annoying to use
Found 2 lint suggestions; failing.
sql/database.go:8:2: a blank import should be only in a main or test package, or have a comment justifying it
Found 1 lint suggestions; failing.
sqlfs/reader.go:64:1: exported method Reader.Close should have comment or be unexported
Found 1 lint suggestions; failing.
sql/alps_submitter.go:127:9: if block ends with a return statement, so drop this else and outdent its block
sql/alps_submitter.go:170:10: if block ends with a return statement, so drop this else and outdent its block
sql/alps_submitter.go:262:13: if block ends with a return statement, so drop this else and outdent its block
sql/alps_submitter.go:293:9: if block ends with a return statement, so drop this else and outdent its block (move short variable declaration to its own line if necessary)
sql/alps_submitter.go:499:9: if block ends with a return statement, so drop this else and outdent its block
Found 5 lint suggestions; failing.
sql/pipe.go:7:5: exported var ErrClosedPipe should have comment or be unexported
Found 1 lint suggestions; failing.
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -27,8 +27,8 @@ install:
 script:
   - ls -lah
   - docker build -t sqlflow:dev -f Dockerfile.dev .
-  - docker run --rm -v $GOPATH:/go -w /go/src/github.com/sql-machine-learning/sqlflow sqlflow:dev pre-commit run -a
-  - docker run --rm -v $GOPATH:/go -w /go/src/github.com/sql-machine-learning/sqlflow sqlflow:dev bash scripts/test.sh
+  - docker run --rm -v $GOPATH/src:/go/src -w /go/src/github.com/sql-machine-learning/sqlflow sqlflow:dev pre-commit run -a
+  - docker run --rm -v $GOPATH/src:/go/src -w /go/src/github.com/sql-machine-learning/sqlflow sqlflow:dev bash scripts/test.sh
 
 deploy:
   skip_cleanup: true

--- a/.travis.yml
+++ b/.travis.yml
@@ -27,6 +27,7 @@ install:
 script:
   - ls -lah
   - docker build -t sqlflow:dev -f Dockerfile.dev .
+  - docker run --rm -v $GOPATH:/go -w /go/src/github.com/sql-machine-learning/sqlflow sqlflow:dev pre-commit run -a
   - docker run --rm -v $GOPATH:/go -w /go/src/github.com/sql-machine-learning/sqlflow sqlflow:dev bash scripts/test.sh
 
 deploy:

--- a/Dockerfile.dev
+++ b/Dockerfile.dev
@@ -29,7 +29,7 @@ RUN /bin/bash -c "source activate sqlflow-dev && python -m pip install \
 RUN wget --quiet https://dl.google.com/go/go1.11.5.linux-amd64.tar.gz
 RUN tar -C /usr/local -xzf go1.11.5.linux-amd64.tar.gz
 RUN rm go1.11.5.linux-amd64.tar.gz
-ENV PATH $PATH:/usr/local/go/bin
+
 
 RUN wget --quiet https://github.com/protocolbuffers/protobuf/releases/download/v3.6.1/protoc-3.6.1-linux-x86_64.zip
 RUN apt-get install -y unzip
@@ -38,6 +38,8 @@ RUN rm protoc-3.6.1-linux-x86_64.zip
 
 RUN mkdir -p /go
 ENV GOPATH /go
+ENV PATH $PATH:/usr/local/go/bin:/go/bin
+
 RUN go get github.com/golang/protobuf/protoc-gen-go && \
 go get -u golang.org/x/lint/golint
 RUN mv /go/bin/protoc-gen-go /usr/local/bin/

--- a/Dockerfile.dev
+++ b/Dockerfile.dev
@@ -38,7 +38,8 @@ RUN rm protoc-3.6.1-linux-x86_64.zip
 
 RUN mkdir -p /go
 ENV GOPATH /go
-RUN go get github.com/golang/protobuf/protoc-gen-go
+RUN go get github.com/golang/protobuf/protoc-gen-go && \
+go get -u golang.org/x/lint/golint
 RUN mv /go/bin/protoc-gen-go /usr/local/bin/
 
 RUN apt-get install -y sqlite3 libsqlite3-dev

--- a/Dockerfile.dev
+++ b/Dockerfile.dev
@@ -22,6 +22,7 @@ RUN /bin/bash -c "source activate sqlflow-dev && python -m pip install \
     tensorflow==2.0.0-alpha0 \
     mysql-connector-python \
     pyhive \
+    pre-commit \
     ${PIP_ADD_PACKAGES} \
     "
 

--- a/server/sqlflowserver.go
+++ b/server/sqlflowserver.go
@@ -1,4 +1,8 @@
-//go:generate protoc -I proto proto/sqlflow.proto --go_out=plugins=grpc:proto
+// Package server is the SQLFlow grpc server which connects to database and
+// parse, submit or execute the training and predicting codes.
+//
+// To generate grpc protobuf code, run the below command:
+// go:generate protoc -I proto proto/sqlflow.proto --go_out=plugins=grpc:proto
 package server
 
 import (
@@ -15,17 +19,18 @@ import (
 )
 
 // NewServer returns a server instance
-func NewServer(run func(string, *sf.DB) *sf.PipeReader, db *sf.DB) *server {
-	return &server{run: run, db: db}
+func NewServer(run func(string, *sf.DB) *sf.PipeReader, db *sf.DB) *Server {
+	return &Server{run: run, db: db}
 }
 
-type server struct {
+// Server is the instance will be used to connect to DB and execute training
+type Server struct {
 	run func(sql string, db *sf.DB) *sf.PipeReader
 	db  *sf.DB
 }
 
 // Run implements `rpc Run (Request) returns (stream Response)`
-func (s *server) Run(req *pb.Request, stream pb.SQLFlow_RunServer) error {
+func (s *Server) Run(req *pb.Request, stream pb.SQLFlow_RunServer) error {
 	pr := s.run(req.Sql, s.db)
 	defer pr.Close()
 

--- a/server/sqlflowserver.go
+++ b/server/sqlflowserver.go
@@ -1,8 +1,9 @@
+//go:generate protoc -I proto proto/sqlflow.proto --go_out=plugins=grpc:proto
+
 // Package server is the SQLFlow grpc server which connects to database and
 // parse, submit or execute the training and predicting codes.
 //
 // To generate grpc protobuf code, run the below command:
-// go:generate protoc -I proto proto/sqlflow.proto --go_out=plugins=grpc:proto
 package server
 
 import (

--- a/server/sqlflowserver_test.go
+++ b/server/sqlflowserver_test.go
@@ -67,7 +67,7 @@ func startServer(done chan bool) {
 
 	s := grpc.NewServer()
 	s.GetServiceInfo()
-	pb.RegisterSQLFlowServer(s, &server{run: mockRun, db: nil})
+	pb.RegisterSQLFlowServer(s, &Server{run: mockRun, db: nil})
 	reflection.Register(s)
 	if err := s.Serve(listener); err != nil {
 		log.Fatalf("failed to serve: %v", err)

--- a/sql/alps_submitter.go
+++ b/sql/alps_submitter.go
@@ -285,10 +285,10 @@ func expression2string(e interface{}) (string, error) {
 	if err != nil {
 		return "", err
 	}
-	if str, ok := resolved.(string); !ok {
-		return "", fmt.Errorf("expression expected to be string, actual: %s", resolved)
+	if str, ok := resolved.(string); ok {
+		return str, nil
 	}
-	return str, nil
+	return "", fmt.Errorf("expression expected to be string, actual: %s", resolved)
 }
 
 func filter(attrs []*attribute, prefix string) []*attribute {

--- a/sql/alps_submitter.go
+++ b/sql/alps_submitter.go
@@ -124,13 +124,12 @@ func (fs *featureSpec) ToString() string {
 			strings.Join(strings.Split(fmt.Sprint(fs.Shape), " "), ","),
 			fs.DType,
 			fs.Delimiter)
-	} else {
-		return fmt.Sprintf("DenseColumn(name=\"%s\", shape=%s, dtype=\"%s\", separator=\"%s\")",
-			fs.FeatureName,
-			strings.Join(strings.Split(fmt.Sprint(fs.Shape), " "), ","),
-			fs.DType,
-			fs.Delimiter)
 	}
+	return fmt.Sprintf("DenseColumn(name=\"%s\", shape=%s, dtype=\"%s\", separator=\"%s\")",
+		fs.FeatureName,
+		strings.Join(strings.Split(fmt.Sprint(fs.Shape), " "), ","),
+		fs.DType,
+		fs.Delimiter)
 }
 
 func resolveFeatureSpec(el *exprlist, isSparse bool) (*featureSpec, error) {
@@ -167,9 +166,8 @@ func resolveExpression(e interface{}) (interface{}, error) {
 	if expr, ok := e.(*expr); ok {
 		if expr.val != "" {
 			return expr.val, nil
-		} else {
-			return resolveExpression(&expr.sexp)
 		}
+		return resolveExpression(&expr.sexp)
 	}
 
 	el, ok := e.(*exprlist)
@@ -259,9 +257,8 @@ func resolveExpression(e interface{}) (interface{}, error) {
 					value, err := resolveExpression(&expr.sexp)
 					if err != nil {
 						return nil, err
-					} else {
-						list = append(list, value)
 					}
+					list = append(list, value)
 				}
 			}
 		}
@@ -290,9 +287,8 @@ func expression2string(e interface{}) (string, error) {
 	}
 	if str, ok := resolved.(string); !ok {
 		return "", fmt.Errorf("expression expected to be string, actual: %s", resolved)
-	} else {
-		return str, nil
 	}
+	return str, nil
 }
 
 func filter(attrs []*attribute, prefix string) []*attribute {
@@ -496,9 +492,8 @@ func trainALPS(wr *PipeWriter, pr *extendedSelect, cwd string) error {
 func submitALPS(w *PipeWriter, pr *extendedSelect, db *DB, cwd string) error {
 	if pr.train {
 		return trainALPS(w, pr, cwd)
-	} else {
-		return fmt.Errorf("inference not supported yet in ALPS")
 	}
+	return fmt.Errorf("inference not supported yet in ALPS")
 }
 
 const alpsTemplateText = `

--- a/sql/database.go
+++ b/sql/database.go
@@ -5,6 +5,7 @@ import (
 	"fmt"
 	"strings"
 
+	// import drivers for heterogonous DB support
 	_ "github.com/go-sql-driver/mysql"
 	_ "github.com/mattn/go-sqlite3"
 	_ "sqlflow.org/gohive"

--- a/sql/pipe.go
+++ b/sql/pipe.go
@@ -4,6 +4,7 @@ import (
 	"errors"
 )
 
+// ErrClosedPipe will occure when manipulating an already closed pipe
 var ErrClosedPipe = errors.New("pipe: write on closed pipe")
 
 // pipe follows the design at https://blog.golang.org/pipelines

--- a/sqlfs/reader.go
+++ b/sqlfs/reader.go
@@ -61,6 +61,7 @@ func (r *Reader) Read(p []byte) (n int, e error) {
 	return n, e
 }
 
+// Close the reader connection to sqlfs
 func (r *Reader) Close() error {
 	if r.rows != nil {
 		if e := r.rows.Close(); e != nil {

--- a/sqlfs/sqlfile_test.go
+++ b/sqlfs/sqlfile_test.go
@@ -65,7 +65,7 @@ func TestWriteAndRead(t *testing.T) {
 	a.Equal(len(buf), n)
 
 	// A big output.
-	buf = make([]byte, kBufSize+1)
+	buf = make([]byte, bufSize+1)
 	for i := range buf {
 		buf[i] = 'x'
 	}
@@ -87,12 +87,12 @@ func TestWriteAndRead(t *testing.T) {
 	a.Equal(2, strings.Count(string(buf), "\n"))
 
 	// A big read of rest
-	buf = make([]byte, kBufSize*2)
+	buf = make([]byte, bufSize*2)
 	n, e = r.Read(buf)
 	a.Equal(io.EOF, e)
-	a.Equal(kBufSize+2, n)
+	a.Equal(bufSize+2, n)
 	a.Equal(1, strings.Count(string(buf), "\n"))
-	a.Equal(kBufSize+1, strings.Count(string(buf), "x"))
+	a.Equal(bufSize+1, strings.Count(string(buf), "x"))
 
 	// Another big read
 	n, e = r.Read(buf)

--- a/sqlfs/writer.go
+++ b/sqlfs/writer.go
@@ -5,7 +5,7 @@ import (
 	"fmt"
 )
 
-var kBufSize = 4 * 1024
+var bufSize = 4 * 1024
 
 // Writer implements io.WriteCloser.
 type Writer struct {
@@ -30,20 +30,21 @@ func Append(db *sql.DB, table string) (*Writer, error) {
 	if e := createTable(db, table); e != nil {
 		return nil, fmt.Errorf("create: %v", e)
 	}
-	return &Writer{db, table, make([]byte, 0, kBufSize), nil}, nil
+	return &Writer{db, table, make([]byte, 0, bufSize), nil}, nil
 }
 
+// Write write bytes to sqlfs and returns (num_bytes, error)
 func (w *Writer) Write(p []byte) (n int, e error) {
 	n = 0
 	for len(p) > 0 {
-		fill := kBufSize - len(w.buf)
+		fill := bufSize - len(w.buf)
 		if fill > len(p) {
 			fill = len(p)
 		}
 		w.buf = append(w.buf, p[:fill]...)
 		p = p[fill:]
 		n += fill
-		if len(w.buf) >= kBufSize {
+		if len(w.buf) >= bufSize {
 			if e = w.flush(); e != nil {
 				return n, fmt.Errorf("writer flush failed: %v", e)
 			}
@@ -52,6 +53,7 @@ func (w *Writer) Write(p []byte) (n int, e error) {
 	return n, nil
 }
 
+// Close the connection of the sqlfs
 func (w *Writer) Close() error {
 	if e := w.flush(); e != nil {
 		return fmt.Errorf("writer flush failed: %v", e)

